R/compat-zeallot.R

```diff
@@ -0,0 +1,34 @@
+# nocov start --- compat-zeallot --- 2020-11-23
+
+# This drop-in file implements a simple version of zeallot::`%<-%`.
+# Please find the most recent version in rlang's repository.
+
+
+`%<-%` <- function(lhs, rhs) {
+  lhs <- substitute(lhs)
+  env <- caller_env()
+
+  if (!is_call(lhs, "c")) {
+    abort("The LHS of `%<-%` must be a `c()` call.")
+  }
+
+  vars <- as.list(lhs[-1])
+
+  if (length(rhs) < length(vars)) {
+    abort("The RHS of `%<-%` must be long enough for the number of assigned variables.")
```

Comment:
Hmmm, that's starting to feel a bit too complex for a shim

Reply:
The advantage of ignoring the other elements is to make it safer in case of extension. E.g. `c(x, y) <- foo()` will keep working if `foo` adds more elements.

Reply:
Hmmmm, I see. That feels like something I'd want to handle explicitly, but then it's maybe again too complicated?

Reply:
Maybe the safe way to use positional matching on the LHS of `%<-%` is to match by names on the RHS like this:

```r
c(expr, env) %<-% foo()[c("expr", "env")]
```

So I'll merge the original implementation that requires equal lengths, and we can iterate from there in the future. Matching by names in the RHS would be a good extension in the future:

```r
# new = old. This means that the non-name part changes meaning as a
# function of the presence of names. There is a similar semantic shift
# in function formals depending on the presence of argument defaults.
c(expr_q = expr, env = env) %<-% foo()

# Equivalent:
c(expr = , env = ) %<-% foo()
c(expr = expr, env = env) %<-% foo()
```